python/lsst/image_cutout_backend/_image_cutout_backend.py

```diff
@@ -0,0 +1,412 @@
+from lsst.daf.base import PropertyList
+from lsst.daf.butler import Butler, DataId, DatasetRef
+from lsst.resources import ResourcePath, ResourcePathExpression
+
+from .projection_finders import ProjectionFinder
+from .stencils import PixelStencil, SkyStencil
+
+
+@dataclasses.dataclass
+class ReadResult:
+    """A struct that aggregates the results of reading a cutout in the image
+    cutout backend.
+    """
+
+    cutout: Union[Image, Mask, MaskedImage, Exposure]
+    """The image cutout itself.
+    """
+
+    sky_stencil: SkyStencil
+    """The original sky-coordinate stencil.
+    """
+
+    pixel_stencil: PixelStencil
+    """A pixel-coordinate representation of the stencil.
+    """
+
+    metadata: PropertyList
+    """Additional FITS metadata about the cutout process.
+
+    This should be merged with ``cutout.getMetadata()`` on write, for types
+    that carry their own metadata.
+    """
+
+    origin_ref: DatasetRef
+    """Fully-resolved reference to the dataset the cutout is from.
+    """
+
+    def mask(self, name: str = "STENCIL") -> None:
+        """Set the bitmask to show the approximate coverage of nonrectangular
+        stencils.
+
+        Parameters
+        ----------
+        name : `str`, optional
+            Name of the mask plane to add and set.
+
+        Notes
+        -----
+        This method modifies `cutout` in place if it is a `Mask`,
+        `MaskedImage`, or `Exposure`.  It does nothing if `cutout` is an
+        `Image`.
+        """
+        if isinstance(self.cutout, Exposure):
+            mask = self.cutout.mask
+        elif isinstance(self.cutout, MaskedImage):
+            mask = self.cutout.mask
+        elif isinstance(self.cutout, Mask):
+            mask = self.cutout
+        else:
+            return
+        mask.addMaskPlane(name)
+        bits = mask.getPlaneBitMask(name)
+        self.pixel_stencil.set_mask(mask, bits)
+
+    def write_fits(self, path: str) -> None:
+        """Write the cutout to a FITS file.
+
+        Parameters
+        ----------
+        path : `str`
+            Local path to the file.
+
+        Notes
+        -----
+        If ``cutout`` is an `Exposure`, this will merge `metadata` into the
+        cutout's attached metadata.  In other cases, `metadata` is written
+        to the primary header without modifying `cutout`.
+        """
+        if isinstance(self.cutout, Exposure):
+            self.cutout.getMetadata().update(self.metadata)
+            self.cutout.writeFits(path)
+        else:
+            self.cutout.writeFits(path, metadata=self.metadata)
+
+
+class ImageCutoutBackend:
+    """High-level interface to the image cutout backend.
+
+    Parameters
+    ----------
+    butler : `Butler`
+        Butler that image cutouts are read from.
+    projection_finder : `ProjectionObject`
+        Object that obtains the WCS and bounding box for butler datasets of
+        different types.  May include caches.
```

Comment:
How is the frontend supposed to create this parameter?

Reply:
For the foreseeable future, `ProjectionFinder.make_default()` will return something that works, and I'm comfortable hard-coding that initially - it's entirely conceivable it will never have to change.

This essentially encodes conventions about what dataset types our pipelines produce and how they are related, and I wasn't comfortable putting that kind of information directly into the `ImageCutoutBackend` class.  But I also wasn't sure if the long term fix is to put these conventions in a configuration file that would be packaged with the service, or to find a way to make this introspectable from the data repository itself.

Reply:
That sounds good. Thanks!